src/dataset/impl_dataset.rs

```diff
@@ -247,13 +247,237 @@ impl<F: Float, D: Data<Elem = F>, I: Dimension> From<ArrayBase<D, I>>
+        indices.shuffle(&mut rng);
+
+        let records = (&self).records().select(Axis(0), &indices);
+        let targets = indices
+            .iter()
+            .map(|x| (self).targets[*x])
+            .collect::<Array1<_>>();
+
+        DatasetBase::new(records, targets)
+    }
+
+    /// Performs K-folding on the dataset.
+    /// The dataset is divided into `k` "fold", each containing
+    /// `(dataset size)/k` samples, used to generate `k` training-validation
+    /// dataset pairs. Each pair contains a validation `Dataset` with `k` samples,
+    ///  the ones contained in the i-th fold, and a training `Dataset` composed by the
+    /// union of all the samples in the remaining folds.
+    ///
+    /// ### Parameters
+    ///
+    /// * `k`: the number of folds to apply
+    ///
+    /// ### Returns
+    ///
+    /// A vector of `k` training-validation Dataset pairs.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// use linfa::dataset::DatasetView;
+    /// use ndarray::{arr1, arr2};
+    ///
+    /// let records = arr2(&[[1.,1.], [2.,1.], [3.,2.], [4.,1.],[5., 3.], [6.,2.]]);
+    /// let targets = arr1(&[1, 1, 0, 1, 0, 0]);
+    ///
+    /// let dataset : DatasetView<f64, usize> = DatasetView::new(records.view(), targets.view());
+    /// let accuracies = dataset.fold(3).into_iter().map(|(train, valid)| {
+    ///     // Here you can train your model and perform validation
+    ///     
+    ///     // let model = params.fit(&dataset);
+    ///     // let predi = model.predict(&valid);
+    ///     // predi.confusion_matrix(&valid).accuracy()  
+    /// });
+    /// ```
+    ///  
+    pub fn fold(&self, k: usize) -> Vec<(Dataset<F, E>, Dataset<F, E>)> {
+        let fold_size = self.targets().len() / k;
+        let mut res = Vec::new();
+
+        // Generates all k folds of records and targets
+        let mut folds_records: Vec<_> = self.records.axis_chunks_iter(Axis(0), fold_size).collect();
+        let mut folds_targets: Vec<_> = self.targets.axis_chunks_iter(Axis(0), fold_size).collect();
+
+        // For each iteration, take the first fold for both records and targets as the validation set and
+        // stack all the other folds to create the training set. In the end, shift both vectors to the left
+        // so that the next fold will appear in the first position in the next iteration
+        for _ in 0..k {
+            let remaining_records = stack(Axis(0), &folds_records.as_slice()[1..]).unwrap();
+            let remaining_targets = stack(Axis(0), &folds_targets.as_slice()[1..]).unwrap();
+
+            res.push((
+                // training
+                Dataset::new(remaining_records, remaining_targets),
+                // validation
+                Dataset::new(folds_records[0].into_owned(), folds_targets[0].into_owned()),
+            ));
+
+            folds_records.rotate_left(1);
+            folds_targets.rotate_left(1);
+        }
+        res
+    }
```

Comment:
How does this look? Is there a better way than shifting the vectors at each iteration to generate all permutations?

Reply:
Maybe collecting into an `Array1` and then using `select` on the chunks indices?

Reply:
yes much better :+1: I did not anticipate that you cannot concatenate views in ndarray
perhaps we can do this without copying by:
 * changing to `fn fold(&'a mut self, k: usize) -> impl Iterator<Item = (Dataset<_, _>, Dataset<_, _>)>`
 * destructing the arrays with `as_slice` and swapping the validation set into place with https://doc.rust-lang.org/stable/std/primitive.slice.html#method.swap_with_slice
 *  return a view with `ArrayView::from_shape`

this would borrow the dataset for the k-folding and return the dataset in original order

Reply:
perhaps we should name that `iter_fold` and add later on

Reply:
Hi, I have a couple of doubts. Please bear with me as this is the first time I've tried to do something like this:

* The way I've understood it is: take a slice for the first fold and a slice for the rest of the array, then, at every step of the iterator, swap the content of the fold with the content of the remaining array at the position of the next fold. Is that what you meant?
* Wouldn't `swap_with_slice` panic if `k` is such that the last chunk does not have the same length as the others?

Reply:
> Hi, I have a couple of doubts. Please bear with me as this is the first time I've tried to do something like this:
> 
>     * The way I've understood it is: take a slice for the first fold and a slice for the rest of the array, then, at every step of the iterator, swap the content of the fold with the content of the remaining array at the position of the next fold. Is that what you meant?
yes and of course swap the content back in the next step otherwise you would change order of chunks

> 
>     * Wouldn't `swap_with_slice` panic if `k` is such that the last chunk does not have the same length as the others?
I would propose to ignore the trail of the dataset and always include it in the training set

> 
>     * Doesn't `swap_with_slice` want a mutable slice?  Can I take a mutable slice from an `ArrayView`? Shouldn't it be only possible for an `ArrayViewMut`?

you should be able to create a mutable view of the underlying arrays with `pub fn view_mut(&mut self) -> ArrayViewMut<'_, A, D> `

Reply:
are you in the middle of creating this or should we merge the PR and amend such a function in a follow-up?

Reply:
> you should be able to create a mutable view of the underlying arrays with pub fn view_mut(&mut self) -> ArrayViewMut<'_, A, D>
So I can only implement it for `Dataset` and not `DatasetView` right? Because I can't take a mutable view of an `ArrayView`.

I'm currently trying to implement it but If you say so we can stop here for this PR and implement it another time.  I still have a commit to push with the changes you suggested. For `fold` i substitute the shifting of the `Vec` with just a swap of the first chunk with the one in the next fold.

Reply:
I've tried my hand at it and I'm having some issues with lifetimes, I think it would be better to leave it for an eventual follow-up PR

Reply:
:+1: